src/main/java/com/faforever/client/login/OAuthValuesReceiver.java

```diff
@@ -0,0 +1,155 @@
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketException;
+import java.net.URI;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/** Opens a minimal HTTP server that retrieves {@literal code} and {@literal state} from the browser. */
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+@RequiredArgsConstructor
+@Slf4j
+public class OAuthValuesReceiver {
+
+  private static final Pattern CODE_PATTERN = Pattern.compile("code=([^ &]+)");
+  private static final Pattern STATE_PATTERN = Pattern.compile("state=([^ &]+)");
+  private static final List<String> ALLOWED_HOSTS = List.of("localhost");
+
+  @VisibleForTesting
+  CompletableFuture<Integer> listenPort = new CompletableFuture<>();
+
+  private final ClientProperties clientProperties;
+  private final PlatformService platformService;
+  private final UserService userService;
+  private final I18n i18n;
+
+  public CompletableFuture<Values> receiveValues(Optional<URI> redirectUri, Optional<OAuthEndpoint> oAuthEndpoint) {
+    return CompletableFuture.supplyAsync(() -> {
+      Collection<URI> redirectUris = getRedirectUris(redirectUri, oAuthEndpoint);
+
+      for (URI uri : redirectUris) {
+        try {
+          return readWithUri(uri);
+        } catch (SocketException e) {
+          log.warn("Port " + uri.getPort() + " is probably already in use", e);
+        }
+      }
+      throw new IllegalStateException("Could not read from any redirect URI: " + redirectUris);
+    });
+  }
+
+  @SneakyThrows
+  private List<URI> getRedirectUris(Optional<URI> redirectUri, Optional<OAuthEndpoint> oAuthEndpoint) {
+    JavaFxUtil.assertBackgroundThread();
+    if (redirectUri.isPresent()) {
+      return Collections.singletonList(redirectUri.get());
+    }
+
+    if (!clientProperties.isUseRemotePreferences()) {
+      throw new NoRedirectUriException("No redirect URI has been specified and remote preferences are disabled");
+    }
+
+    OAuthEndpoint endpoint = oAuthEndpoint.orElseThrow(() -> new IllegalStateException("No endpoint has been specified"));
+    List<URI> redirectUris = endpoint.getRedirectUris().stream()
+        .filter(uri -> uri.getPort() > 0)
+        .filter(uri -> ALLOWED_HOSTS.contains(uri.getHost()))
+        .collect(Collectors.toList());
+
+    if (redirectUris.isEmpty()) {
+      throw new NoRedirectUriException("No valid redirect URI has been provided by endpoint: " + endpoint.getUrl());
+    }
+
+    return redirectUris;
+  }
+
+  @SneakyThrows
+  private Values readWithUri(URI uri) throws SocketException {
```

Comment:
What is throwing the SocketException and why not just throw the IOException and catch that? Then we can remove the SneakyThrows and has basically the same effect.

Reply:
Trying to open a socket on a port that is already occupied throws a SocketException, which is an IOException. If I throw IOException, the caller is forced to handle which is undesired because it only wants to handle SocketException.

Reply:
I would still think it might be better to also handle the IOexception should that happen. But since it is inside a completableFuture call then I guess it is okay.

Reply:
Most of my thinking or reasoning is that especially on these more network based things I have had too many experiences of sneaky throws just swallowing connection issues and the like and then the user is just left stranded.

Although hopefully in this case it is caught by the exceptionally.

Reply:
Checked exceptions were arguably a bad idea, which is also why they don't exist e.g. in Kotlin. @SneakyThrows just kind of retro-fits this so you might want to rethink your position ;-)

In case of a blocked port, I need to continue looking for another port. In case of any other IOException I don't want to keep looking but let the exception propagate.

Reply:
Yep understood they are just different bandaids over a Library problem

Reply:
If it wasn't called primarily from a completable future I would say wrap it in one of the notifiable exceptions but unfortunately those don't work inside completable futures.